demos/pkey/EVP_PKEY_EC_keygen.c

```diff
@@ -0,0 +1,168 @@
+
+/*
+ * The following code shows how to generate a EC key from a curve name
+ * with additional parameters. If only the curve name is required then the
+ * simple helper can be used instead i.e. Either
+ * pkey = EVP_EC_gen(curvename); OR
+ * pkey = EVP_PKEY_Q_keygen(libctx, propq, "EC", curvename);
+ */
+static EVP_PKEY *do_ec_keygen(void)
+{
+    /*
+     * The libctx and propq can be set if required, they are included here
+     * to show how they are passed to EVP_PKEY_CTX_new_from_name().
+     */
+    OSSL_LIB_CTX *libctx = NULL;
+    const char *propq = NULL;
+    static int use_cofactordh = 1;
+    EVP_PKEY *key = NULL;
+    OSSL_PARAM params[3];
+    EVP_PKEY_CTX *genctx = NULL;
+    static const char *curvename = "P-256";
+
+    genctx = EVP_PKEY_CTX_new_from_name(libctx, "EC", propq);
+    if (genctx == NULL) {
+        fprintf(stderr, "EVP_PKEY_CTX_new_from_name() failed\n");
+        goto cleanup;
+    }
+
+    if (EVP_PKEY_keygen_init(genctx) <= 0) {
+        fprintf(stderr, "EVP_PKEY_keygen_init() failed\n");
+        goto cleanup;
+    }
+
+    params[0] = OSSL_PARAM_construct_utf8_string(OSSL_PKEY_PARAM_GROUP_NAME,
+                                                 (char *)curvename, 0);
+    /*
+     * This is an optional parameter.
+     * For many curves the cofactor is 1 - so setting this has
+     * no effect.
+     */
+    params[1] = OSSL_PARAM_construct_int(OSSL_PKEY_PARAM_USE_COFACTOR_ECDH,
+                                         &use_cofactordh);
+    params[2] = OSSL_PARAM_construct_end();
+    if (!EVP_PKEY_CTX_set_params(genctx, params)) {
+        fprintf(stderr, "EVP_PKEY_CTX_set_params() failed\n");
+        goto cleanup;
+    }
+
+    fprintf(stdout, "Generating EC key\n\n");
+    if (EVP_PKEY_generate(genctx, &key) <= 0) {
+        fprintf(stderr, "EVP_PKEY_generate() failed\n");
+        goto cleanup;
+    }
+cleanup:
+    EVP_PKEY_CTX_free(genctx);
+    return key;
+}
+
+/*
+ * The following code shows how retrieve key data from the generated
+ * EC key. See doc/man7/EVP_PKEY-EC.pod for more information.
+ *
+ * EVP_PKEY_print_private() could also be used to display the values.
+ */
+static int get_key_values(EVP_PKEY *pkey)
+{
+    int result = 0;
+    char out_curvename[80];
+    unsigned char out_pubkey[80];
+    char out_pub_hexstr[256];
+    char *out_priv_hexstr = NULL;
+    BIGNUM *out_priv = NULL;
+    size_t i, len = 0;
+
+    if (!EVP_PKEY_get_utf8_string_param(pkey, OSSL_PKEY_PARAM_GROUP_NAME,
+                                        out_curvename, sizeof(out_curvename),
+                                        NULL)) {
+        fprintf(stderr, "Failed to get curve name\n");
+        goto cleanup;
+    }
+
+    if (!EVP_PKEY_get_octet_string_param(pkey, OSSL_PKEY_PARAM_PUB_KEY,
+                                        out_pubkey, sizeof(out_pubkey), &len)) {
+        fprintf(stderr, "Failed to get public key\n");
+        goto cleanup;
+    }
+
+    out_priv = BN_new();
```

Comment:
Doesn't `EVP_PKEY_get_bn_param()` do the allocation for us?  I.e. this needn't be done separately?

Reply:
I dont think so..

Reply:
If it calls down to the OSSL_PARAM_get_BN function it will (assuming the pointer is initialised to NULL).